File: apps/api/src/opensolve_pipe/models/project.py
"""Project container and metadata models."""

from __future__ import annotations

from datetime import datetime
from typing import TYPE_CHECKING
from uuid import uuid4

from pydantic import Field, field_validator, model_validator

from .base import OpenSolvePipeBaseModel
from .components import Component, PumpComponent
from .connections import (
    PipeConnection,
    validate_port_direction_compatibility,
)
from .fluids import FluidDefinition
from .units import SolverOptions, UnitPreferences

if TYPE_CHECKING:
    from .pump import PumpCurve
    from .results import SolvedState


class ProjectMetadata(OpenSolvePipeBaseModel):
    """Project metadata and versioning info."""

    name: str = Field(default="Untitled Project", description="Project name")
    description: str | None = Field(default=None, description="Project description")
    author: str | None = Field(default=None, description="Project author")
    created: datetime = Field(
        default_factory=datetime.utcnow, description="Creation timestamp"
    )
    modified: datetime = Field(
        default_factory=datetime.utcnow, description="Last modification timestamp"
    )
    version: str = Field(default="1", description="Project version")
    parent_version: str | None = Field(
        default=None, description="Parent version for branching"
    )


class ProjectSettings(OpenSolvePipeBaseModel):
    """Project-level settings."""

    units: UnitPreferences = Field(
        default_factory=UnitPreferences, description="Unit preferences"
    )
    enabled_checks: list[str] = Field(
        default_factory=list, description="Enabled design checks"
    )
    solver_options: SolverOptions = Field(
        default_factory=SolverOptions, description="Solver configuration"
    )


class Project(OpenSolvePipeBaseModel):
    """Top-level project container."""

    id: str = Field(
        default_factory=lambda: str(uuid4()), description="Unique project identifier"
    )
    metadata: ProjectMetadata = Field(
        default_factory=ProjectMetadata, description="Project metadata"
    )
    settings: ProjectSettings = Field(
        default_factory=ProjectSettings, description="Project settings"
    )
    fluid: FluidDefinition = Field(
        default_factory=FluidDefinition, description="Working fluid definition"
    )
    components: list[Component] = Field(
        default_factory=list, description="Network components"
    )
    connections: list[PipeConnection] = Field(
        default_factory=list,
        description="Port-based pipe connections between components",
    )
    pump_library: list[PumpCurve] = Field(
        default_factory=list, description="Available pump curves"
    )
    results: SolvedState | None = Field(
        default=None, description="Solved state (if network has been solved)"
    )

    @field_validator("components")
    @classmethod
    def validate_component_ids_unique(cls, v: list[Component]) -> list[Component]:
        """Ensure all component IDs are unique."""
        ids = [c.id for c in v]
        if len(ids) != len(set(ids)):
            duplicates = [id for id in ids if ids.count(id) > 1]
            raise ValueError(f"Duplicate component IDs: {set(duplicates)}")
        return v

    @field_validator("components")
    @classmethod
    def validate_component_references(cls, v: list[Component]) -> list[Component]:
        """Ensure all downstream connection references are valid."""
        component_ids = {c.id for c in v}
        for component in v:
            for conn in component.downstream_connections:
                if conn.target_component_id not in component_ids:
                    raise ValueError(
                        f"Component '{component.id}' references unknown "
                        f"target: '{conn.target_component_id}'"
                    )
        return v

    @field_validator("pump_library")
    @classmethod
    def validate_pump_curve_ids_unique(cls, v: list[PumpCurve]) -> list[PumpCurve]:
        """Ensure all pump curve IDs are unique."""
        ids = [c.id for c in v]
        if len(ids) != len(set(ids)):
            duplicates = [id for id in ids if ids.count(id) > 1]
            raise ValueError(f"Duplicate pump curve IDs: {set(duplicates)}")
        return v

    @model_validator(mode="after")
    def validate_pump_curve_references(self) -> Project:
        """Ensure pump components reference valid pump curves."""
        curve_ids = {c.id for c in self.pump_library}
        for component in self.components:
            if (
                isinstance(component, PumpComponent)
                and component.curve_id not in curve_ids
            ):
                raise ValueError(
                    f"Pump '{component.id}' references unknown "
                    f"curve: '{component.curve_id}'"
                )
        return self

    @field_validator("connections")
    @classmethod
    def validate_connection_ids_unique(
        cls, v: list[PipeConnection]
    ) -> list[PipeConnection]:
        """Ensure all connection IDs are unique."""
        ids = [c.id for c in v]
        if len(ids) != len(set(ids)):
            duplicates = [id for id in ids if ids.count(id) > 1]
            raise ValueError(f"Duplicate connection IDs: {set(duplicates)}")
        return v

    @model_validator(mode="after")
    def validate_connection_references(self) -> Project:
        """Validate that all connections reference valid components and ports."""
        component_map = {c.id: c for c in self.components}

        for conn in self.connections:
            # Validate from component exists
            if conn.from_component_id not in component_map:
                raise ValueError(
                    f"Connection '{conn.id}' references unknown "
                    f"source component: '{conn.from_component_id}'"
                )

            # Validate to component exists
            if conn.to_component_id not in component_map:
                raise ValueError(
                    f"Connection '{conn.id}' references unknown "
                    f"target component: '{conn.to_component_id}'"
                )

            # Validate from port exists
            from_component = component_map[conn.from_component_id]
            from_port = from_component.get_port(conn.from_port_id)
            if from_port is None:
                raise ValueError(
                    f"Connection '{conn.id}' references unknown port "
                    f"'{conn.from_port_id}' on component '{conn.from_component_id}'"
                )

            # Validate to port exists
            to_component = component_map[conn.to_component_id]
            to_port = to_component.get_port(conn.to_port_id)
            if to_port is None:
                raise ValueError(
                    f"Connection '{conn.id}' references unknown port "
                    f"'{conn.to_port_id}' on component '{conn.to_component_id}'"
                )

            # Validate port direction compatibility
            dir_valid, dir_error = validate_port_direction_compatibility(
                from_port, to_port
            )
            if not dir_valid:
                raise ValueError(f"Connection '{conn.id}': {dir_error}")

        return self

    def get_component(self, component_id: str) -> Component | None:
        """Get a component by ID."""
        for component in self.components:
            if component.id == component_id:
                return component
        return None

    def get_pump_curve(self, curve_id: str) -> PumpCurve | None:
        """Get a pump curve by ID."""
        for curve in self.pump_library:
            if curve.id == curve_id:
                return curve
        return None

    def get_connection(self, connection_id: str) -> PipeConnection | None:
        """Get a connection by ID."""
        for conn in self.connections:
            if conn.id == connection_id:
                return conn
        return None

    def get_connections_from_component(self, component_id: str) -> list[PipeConnection]:
        """Get all connections originating from a component."""
        return [c for c in self.connections if c.from_component_id == component_id]

    def get_connections_to_component(self, component_id: str) -> list[PipeConnection]:
        """Get all connections targeting a component."""
        return [c for c in self.connections if c.to_component_id == component_id]

    def get_connections_for_port(
        self, component_id: str, port_id: str
    ) -> list[PipeConnection]:
        """Get all connections involving a specific port."""
        return [
            c
            for c in self.connections
            if (c.from_component_id == component_id and c.from_port_id == port_id)
            or (c.to_component_id == component_id and c.to_port_id == port_id)
        ]

File: apps/web/src/lib/models/project.ts
/**
 * Project container and metadata models.
 * Mirrors: apps/api/src/opensolve_pipe/models/project.py
 */

import type { Component, PipeConnection } from './components';
import type { FluidDefinition } from './fluids';
import { DEFAULT_FLUID_DEFINITION } from './fluids';
import type { PumpCurve } from './pump';
import type { SolvedState } from './results';
import type { SolverOptions, UnitPreferences } from './units';
import { DEFAULT_SOLVER_OPTIONS, DEFAULT_UNIT_PREFERENCES } from './units';

/** Project metadata and versioning info. */
export interface ProjectMetadata {
	/** Project name. Default: "Untitled Project" */
	name: string;
	/** Project description. */
	description?: string;
	/** Project author. */
	author?: string;
	/** Creation timestamp (ISO 8601 string). */
	created: string;
	/** Last modification timestamp (ISO 8601 string). */
	modified: string;
	/** Project version. Default: "1" */
	version: string;
	/** Parent version for branching. */
	parent_version?: string;
}

/** Project-level settings. */
export interface ProjectSettings {
	/** Unit preferences. */
	units: UnitPreferences;
	/** Enabled design checks. */
	enabled_checks: string[];
	/** Solver configuration. */
	solver_options: SolverOptions;
}

/** Top-level project container. */
export interface Project {
	/** Unique project identifier. */
	id: string;
	/** Project metadata. */
	metadata: ProjectMetadata;
	/** Project settings. */
	settings: ProjectSettings;
	/** Working fluid definition. */
	fluid: FluidDefinition;
	/** Network components. */
	components: Component[];
	/** Port-based connections between components. */
	connections: PipeConnection[];
	/** Available pump curves. */
	pump_library: PumpCurve[];
	/** Solved state (if network has been solved). */
	results?: SolvedState;
}

/** Generate a unique project ID. */
export function generateProjectId(): string {
	return `proj_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
}

/** Create default project metadata. */
export function createDefaultMetadata(): ProjectMetadata {
	const now = new Date().toISOString();
	return {
		name: 'Untitled Project',
		created: now,
		modified: now,
		version: '1'
	};
}

/** Create default project settings. */
export function createDefaultSettings(): ProjectSettings {
	return {
		units: { ...DEFAULT_UNIT_PREFERENCES },
		enabled_checks: [],
		solver_options: { ...DEFAULT_SOLVER_OPTIONS }
	};
}

/** Create a new empty project. */
export function createNewProject(): Project {
	return {
		id: generateProjectId(),
		metadata: createDefaultMetadata(),
		settings: createDefaultSettings(),
		fluid: { ...DEFAULT_FLUID_DEFINITION },
		components: [],
		connections: [],
		pump_library: []
	};
}

/** Get a component by ID from a project. */
export function getComponentById(project: Project, componentId: string): Component | undefined {
	return project.components.find((c) => c.id === componentId);
}

/** Get a pump curve by ID from a project. */
export function getPumpCurveById(project: Project, curveId: string): PumpCurve | undefined {
	return project.pump_library.find((c) => c.id === curveId);
}

/** Check if all component IDs are unique. */
export function validateComponentIds(project: Project): string | null {
	const ids = project.components.map((c) => c.id);
	const uniqueIds = new Set(ids);
	if (ids.length !== uniqueIds.size) {
		const duplicates = ids.filter((id, index) => ids.indexOf(id) !== index);
		return `Duplicate component IDs: ${[...new Set(duplicates)].join(', ')}`;
	}
	return null;
}

/** Check if all downstream connection references are valid. */
export function validateConnectionReferences(project: Project): string | null {
	const componentIds = new Set(project.components.map((c) => c.id));
	for (const component of project.components) {
		for (const conn of component.downstream_connections) {
			if (!componentIds.has(conn.target_component_id)) {
				return `Component '${component.id}' references unknown target: '${conn.target_component_id}'`;
			}
		}
	}
	return null;
}

/** Check if all pump curve IDs are unique. */
export function validatePumpCurveIds(project: Project): string | null {
	const ids = project.pump_library.map((c) => c.id);
	const uniqueIds = new Set(ids);
	if (ids.length !== uniqueIds.size) {
		const duplicates = ids.filter((id, index) => ids.indexOf(id) !== index);
		return `Duplicate pump curve IDs: ${[...new Set(duplicates)].join(', ')}`;
	}
	return null;
}

/** Check if all pump components reference valid pump curves. */
export function validatePumpCurveReferences(project: Project): string | null {
	const curveIds = new Set(project.pump_library.map((c) => c.id));
	for (const component of project.components) {
		if (component.type === 'pump' && !curveIds.has(component.curve_id)) {
			return `Pump '${component.id}' references unknown curve: '${component.curve_id}'`;
		}
	}
	return null;
}

/** Check if all pipe connection IDs are unique. */
export function validatePipeConnectionIds(project: Project): string | null {
	const ids = project.connections.map((c) => c.id);
	const uniqueIds = new Set(ids);
	if (ids.length !== uniqueIds.size) {
		const duplicates = ids.filter((id, index) => ids.indexOf(id) !== index);
		return `Duplicate connection IDs: ${[...new Set(duplicates)].join(', ')}`;
	}
	return null;
}

/** Check if all pipe connection references are valid. */
export function validatePipeConnectionReferences(project: Project): string | null {
	const componentMap = new Map(project.components.map((c) => [c.id, c]));

	for (const conn of project.connections) {
		const fromComponent = componentMap.get(conn.from_component_id);
		if (!fromComponent) {
			return `Connection '${conn.id}' references unknown source component: '${conn.from_component_id}'`;
		}
		if (!fromComponent.ports.some((p) => p.id === conn.from_port_id)) {
			return `Connection '${conn.id}' references unknown source port: '${conn.from_port_id}' on component '${conn.from_component_id}'`;
		}

		const toComponent = componentMap.get(conn.to_component_id);
		if (!toComponent) {
			return `Connection '${conn.id}' references unknown target component: '${conn.to_component_id}'`;
		}
		if (!toComponent.ports.some((p) => p.id === conn.to_port_id)) {
			return `Connection '${conn.id}' references unknown target port: '${conn.to_port_id}' on component '${conn.to_component_id}'`;
		}
	}
	return null;
}

/** Validate entire project. Returns array of error messages (empty if valid). */
export function validateProject(project: Project): string[] {
	const errors: string[] = [];

	const componentIdError = validateComponentIds(project);
	if (componentIdError) errors.push(componentIdError);

	const connectionError = validateConnectionReferences(project);
	if (connectionError) errors.push(connectionError);

	const curveIdError = validatePumpCurveIds(project);
	if (curveIdError) errors.push(curveIdError);

	const curveRefError = validatePumpCurveReferences(project);
	if (curveRefError) errors.push(curveRefError);

	const pipeConnIdError = validatePipeConnectionIds(project);
	if (pipeConnIdError) errors.push(pipeConnIdError);

	const pipeConnRefError = validatePipeConnectionReferences(project);
	if (pipeConnRefError) errors.push(pipeConnRefError);

	return errors;
}

/** Update project modified timestamp. */
export function touchProject(project: Project): Project {
	return {
		...project,
		metadata: {
			...project.metadata,
			modified: new Date().toISOString()
		}
	};
}

/** Create a new version (branch) of a project. */
export function branchProject(project: Project, newName?: string): Project {
	const now = new Date().toISOString();
	const newVersion = String(parseInt(project.metadata.version) + 1);

	return {
		...project,
		id: generateProjectId(),
		metadata: {
			...project.metadata,
			name: newName ?? `${project.metadata.name} (copy)`,
			created: now,
			modified: now,
			version: newVersion,
			parent_version: project.metadata.version
		},
		connections: [...project.connections], // Deep copy connections
		results: undefined // Clear results for new version
	};
}

/** Generate a unique connection ID. */
export function generateConnectionId(): string {
	return `conn_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
}

/** Get a connection by ID from a project. */
export function getConnectionById(project: Project, connectionId: string): PipeConnection | undefined {
	return project.connections.find((c) => c.id === connectionId);
}

/** Get all connections from a component. */
export function getConnectionsFromComponent(project: Project, componentId: string): PipeConnection[] {
	return project.connections.filter((c) => c.from_component_id === componentId);
}

/** Get all connections to a component. */
export function getConnectionsToComponent(project: Project, componentId: string): PipeConnection[] {
	return project.connections.filter((c) => c.to_component_id === componentId);
}

/** Get all connections involving a specific port. */
export function getConnectionsForPort(
	project: Project,
	componentId: string,
	portId: string
): PipeConnection[] {
	return project.connections.filter(
		(c) =>
			(c.from_component_id === componentId && c.from_port_id === portId) ||
			(c.to_component_id === componentId && c.to_port_id === portId)
	);
}
